File: apps/ios/Anky/Features/Map/MapViewModel.swift
import Foundation

@MainActor
final class MapViewModel: ObservableObject {
    @Published private(set) var days: [SessionDay] = []
    @Published private(set) var spatialDays: [SessionDay] = []
    @Published private(set) var firstOpenDate: Date = Date()
    @Published private(set) var mapStartDate: Date = Date()
    @Published private(set) var todayDate: Date = Calendar.ankyUTC.startOfDay(for: Date())
    @Published private(set) var completeAnkyCount = 0
    @Published private(set) var totalWritingMinutes = 0
    @Published private(set) var currentStreak = 0
    @Published private(set) var completeAnkySessions: [SessionSummary] = []

    private let archive: LocalAnkyArchive
    private let reflectionStore: ReflectionStore
    private let sessionIndexStore: SessionIndexStore
    private let appOpenStore: AppOpenStore
    private let calendar: Calendar

    init(
        archive: LocalAnkyArchive = LocalAnkyArchive(),
        reflectionStore: ReflectionStore = ReflectionStore(),
        sessionIndexStore: SessionIndexStore = SessionIndexStore(),
        appOpenStore: AppOpenStore = AppOpenStore(),
        calendar: Calendar = .ankyUTC
    ) {
        self.archive = archive
        self.reflectionStore = reflectionStore
        self.sessionIndexStore = sessionIndexStore
        self.appOpenStore = appOpenStore
        self.calendar = calendar
        refresh()
    }

    func refresh() {
        let sessions = (try? sessionIndexStore.rebuild(archive: archive, reflectionStore: reflectionStore)) ?? sessionIndexStore.load()
        let now = Date()
        todayDate = calendar.startOfDay(for: now)
        firstOpenDate = JourneyAnchor.anchorDay(in: sessions, fallback: now, calendar: calendar)
        mapStartDate = firstOpenDate
        days = sessions.groupedByDay(calendar: calendar, firstOpenDate: mapStartDate, now: now)
        spatialDays = sessions.groupedByContinuousDays(calendar: calendar, firstOpenDate: mapStartDate, now: now)
        let completeSessions = sessions.filter(\.isComplete)
        completeAnkySessions = completeSessions.sorted { $0.createdAt > $1.createdAt }
        completeAnkyCount = completeSessions.count
        let totalDurationMs = sessions.reduce(Int64(0)) { $0 + $1.durationMs }
        totalWritingMinutes = sessions.isEmpty ? 0 : max(1, Int((totalDurationMs + 59_999) / 60_000))
        currentStreak = Self.currentStreak(from: completeSessions.map(\.createdAt), calendar: calendar, now: now)
    }

    var today: SessionDay? {
        spatialDays.first { calendar.isDate($0.date, inSameDayAs: todayDate) }
    }

    var trailDays: [SessionDay] {
        days.filter { !calendar.isDate($0.date, inSameDayAs: todayDate) }
    }

    func day(for date: Date) -> SessionDay? {
        spatialDays.first { calendar.isDate($0.date, inSameDayAs: date) }
    }

    func artifact(for summary: SessionSummary) -> SavedAnky? {
        try? archive.load(url: summary.localFileURL)
    }

    private static func currentStreak(from dates: [Date], calendar: Calendar, now: Date) -> Int {
        let activeDays = Set(dates.map { calendar.startOfDay(for: $0) })
        guard !activeDays.isEmpty else {
            return 0
        }

        var day = calendar.startOfDay(for: now)
        guard activeDays.contains(day) else {
            return 0
        }

        var streak = 0
        while activeDays.contains(day) {
            streak += 1
            guard let previousDay = calendar.date(byAdding: .day, value: -1, to: day) else {
                break
            }
            day = previousDay
        }
        return streak
    }
}

File: apps/android/app/src/main/java/inc/anky/android/feature/map/MapViewModel.kt
package inc.anky.android.feature.map

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import inc.anky.android.core.storage.AppOpenStore
import inc.anky.android.core.storage.LocalAnkyArchive
import inc.anky.android.core.storage.ReflectionStore
import inc.anky.android.core.storage.SavedAnky
import inc.anky.android.core.storage.SessionDay
import inc.anky.android.core.storage.SessionIndexStore
import inc.anky.android.core.storage.SessionSummary
import java.time.Instant
import java.time.ZoneOffset
import java.time.temporal.ChronoUnit
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

data class MapState(
    val days: List<SessionDay> = emptyList(),
    val completeAnkySessions: List<SessionSummary> = emptyList(),
    val completeAnkyCount: Int = 0,
    val totalWritingMinutes: Int = 0,
    val currentStreak: Int = 0,
    val errorMessage: String? = null,
)

class MapViewModel(
    private val archive: LocalAnkyArchive,
    private val reflectionStore: ReflectionStore,
    private val indexStore: SessionIndexStore,
    private val appOpenStore: AppOpenStore,
    private val archiveDispatcher: CoroutineDispatcher = Dispatchers.IO,
) : ViewModel() {
    private val _state = MutableStateFlow(MapState())
    val state: StateFlow<MapState> = _state

    /**
     * Every saved writing, newest first, for the archive chamber. Loaded
     * off-main (reconstructing text is not free) and searched through the
     * lazily built lowercase cache in [ArchiveSearchIndex].
     */
    private val _archiveEntries = MutableStateFlow<List<SavedAnky>>(emptyList())
    val archiveEntries: StateFlow<List<SavedAnky>> = _archiveEntries

    fun refreshArchiveEntries() {
        viewModelScope.launch(archiveDispatcher) {
            val loaded = runCatching { archive.list() }.getOrDefault(emptyList())
            _archiveEntries.value = loaded.sortedByDescending { it.createdAt }
        }
    }

    fun refresh() {
        runCatching {
            val storedFirstOpen = appOpenStore.loadOrCreate()
            val sessions = runCatching {
                indexStore.rebuild(archive, reflectionStore)
            }.getOrElse {
                indexStore.load()
            }
            val firstOpen = sessions
                .minByOrNull { it.createdAt }
                ?.let { appOpenStore.recordEarlierFirstOpenDate(it.createdAt) }
                ?: storedFirstOpen
            val days = SessionIndexStore.groupByContinuousDays(sessions, firstOpen)
            val completeSessions = sessions.filter { it.isComplete }
            val totalDurationMs = sessions.sumOf { it.durationMs }
            MapState(
                days = days,
                completeAnkySessions = completeSessions.sortedByDescending { it.createdAt },
                completeAnkyCount = completeSessions.size,
                totalWritingMinutes = if (sessions.isEmpty()) 0 else maxOf(1, ((totalDurationMs + 59_999L) / 60_000L).toInt()),
                currentStreak = currentStreak(completeSessions.map { it.createdAt }),
                errorMessage = null,
            )
        }.onSuccess { mapState ->
            _state.value = mapState
        }.onFailure {
            _state.value = MapState(
                days = emptyList(),
                errorMessage = "Could not load the map.",
            )
        }
    }

    private fun currentStreak(dates: List<Instant>, now: Instant = Instant.now()): Int {
        val activeDays = dates.map { it.atZone(ZoneOffset.UTC).toLocalDate() }.toSet()
        if (activeDays.isEmpty()) return 0
        var day = now.atZone(ZoneOffset.UTC).toLocalDate()
        if (!activeDays.contains(day)) return 0
        var streak = 0
        while (activeDays.contains(day)) {
            streak += 1
            day = day.minus(1, ChronoUnit.DAYS)
        }
        return streak
    }
}
